crates/shared/src/price_estimation/trade_verifier.rs

```diff
@@ -179,37 +201,75 @@ impl TradeVerifying for TradeVerifier {
                 .context("could not compute out_amount")?,
             gas: summary.gas_used.as_u64(),
             solver: trade.solver,
+            buy_token_buffer: summary.buy_token_buffer,
         };
         tracing::debug!(
             ?query,
             ?verification,
             promised_gas = trade.gas_estimate,
             promised_out_amount =? trade.out_amount,
             ?verified,
+            internalized = internalize,
             "verified quote"
         );
         Ok(verified)
     }
 }
 
+#[async_trait::async_trait]
+impl TradeVerifying for TradeVerifier {
+    async fn verify(
+        &self,
+        query: &PriceQuery,
+        verification: &Verification,
+        trade: Trade,
+    ) -> Result<VerifiedEstimate> {
+        let regular = self
+            .verify_inner(query, verification, &trade, false)
+            .await?;
+        let buy_amount = match query.kind {
+            OrderKind::Sell => regular.out_amount,
+            OrderKind::Buy => query.in_amount.get(),
+        };
+
+        let sell_token_trusted = self.trusted_tokens.contains(&query.sell_token);
+        if !sell_token_trusted || regular.buy_token_buffer < buy_amount {
+            // Internalized buffer trade is forbidden or will fail.
+            return Ok(regular);
+        }
+
+        // Compute quote based on internal buffer trading.
+        let Ok(internalized) = self.verify_inner(query, verification, &trade, true).await else {
```

Comment:
I don't think the autopilot should decide whether or not to simulate with internalized interactions or not on a case by case basis (but instead leave this logic up to the solvers and simply tell them the settlement contract balance and trusted tokens like for regular solutions). 

The only question IMO should be if we simulate the internalized or uninternalized calldata (I agree with you it's good to do the former to show more competitive prices).

Reply:
Seems like a reasonable solution which will probably result in the `/solve` logic to align closer with the `/quote` logic.

> Maybe I'm understanding things wrongly, but I think we should simply treat the solvers response like the response they will later send on /solve and simulate it exactly the same way we will later do in the driver/autopilot.

That would be the easiest solution but this would also require things like approvals and `ETH` unwraps to happen before a user can even get a verified quote.
These are not a concerns when simulating actual settlements because there we only want to know whether a settlement works right now as is and not theoretically under some additional conditions.

Reply:
> Also the user didn't sign any order that could be executed during a regular `eth_call`.

Maybe I'm not fully understanding how verified quotes work on a technical level, but in order to simulate a user having signed their order, I think we should be able to use `StorageAccessible.simulateDelegatecall` [[1](https://github.com/cowprotocol/contracts/blob/main/src/contracts/mixins/StorageAccessible.sol)] to simulate having set the presignature of a given order ID (making sure that the raw settlement transaction should pass)

Reply:
I explained the idea in detail in the original [PR](https://github.com/cowprotocol/services/pull/1575). The TL;DR is that we use state overrides to put a helper contract at the trader's `from` address. With the trader being turned into a contract we control we mainly use it for 3 things:
* make it do stuff the user would normally have to do to prepare the trade (e.g. wrapping and approvals)
* fake the signature by always returning true from `isValidSignature()`
* proxy any unexpected function call to the original code at the `from` address (in case it was actually a smart contract)

So while your suggestion might be used for faking a signature I think we'd still have the problem of the missing approvals and unwraps. And if we have to impersonate the trader anyway I think we might as well keep the `isValidSignature()` trickery because it's trivial by then.

That being said I have very little experience with smart contracts so it could very well be that there is a better solution for this.